radsat-sk/src/tasks/RCommunicationTasks.c

```diff
@@ -0,0 +1,441 @@
+					// TODO: determine if this message was signalling the end of telecommand mode
+					response_t response = responseACK;
+					int endOfTelecommandMode = 1;
+
+					// prepare for file transfer mode (if necessary)
+					if (endOfTelecommandMode)
+						state.mode = commModeFileTransfer;
+
+					// prepare to send ACK/NACK response
+					state.telecommand.responseToSend = response;
+					state.telecommand.transmitReady = responseStateReady;
+				}
+
+				// file transfer mode, awaiting ACK/NACK from the Ground Station
+				else if (state.mode == commModeFileTransfer && !state.fileTransfer.transmitReady)
+				{
+					// TODO: forward message to command centre and extract the received ACK/NACK response
+					response_t response = responseACK;
+
+					// prepare to send subsequent (or resend previous) file transfer frame
+					state.fileTransfer.responseReceived = response;
+					state.fileTransfer.transmitReady = responseStateReady;
+				}
+			}
+		}
+
+		vTaskDelay(1);
+	}
+}
+
+
+/**
+ * Transmits outgoing frames to the Transceiver's transmitter buffer.
+ *
+ * When NOT in a pass mode, this task does nothing. When in a pass and in the telecommand mode
+ * (i.e. receiving telecommands), this task is responsible for transmitting the appropriate ACKs
+ * (or NACKs) and updating the flags (that are local and private to this module). When in a pass
+ * and in the File Transfer mode, this task is responsible for transmitting frames that are ready
+ * for downlink transmission. Preperation of downlink messages is done by another module prior to
+ * the pass duration.
+ *
+ * @note	This is a high priority task.
+ * @note	When an operational error occurs (e.g. a call to the transceiver module failed), this
+ * 			Task will simply ignore the operation and try again next time. Lower level modules
+ * 			(e.g. the Transceiver) are responsible for reporting these errors to the system.
+ * @param	parameters Unused.
+ */
+void communicationTxTask(void* parameters) {
+	(void)parameters;
+
+	int error = 0;												// error detection
+	uint8_t txSlotsRemaining = TRANCEIVER_TX_MAX_FRAME_COUNT;	// number of open frame slots in the transmitter's buffer
+	uint8_t txMessageSize = 0;									// size (in bytes) of an outgoing frame
+	uint8_t txMessage[TRANCEIVER_TX_MAX_FRAME_SIZE] = { 0 };	// output buffer for messages to be transmitted
+
+	while (1) {
+
+		// pass mode is active
+		if (communicationPassModeActive()) {
+
+			// ready to send ACK/NACK to the Ground Station
+			if (state.telecommand.transmitReady) {
+
+				// serialize the ACK/NACK response to be sent
+				// TODO: use real functions
+//				if (state.telecommand.responseToSend == responseACK)
+//					downlinkSerializedAck(txMessage, &txMessageSize);
+//				else
+//					downlinkSerializedNack(txMessage, &txMessageSize);
+
+				// send the message
+				error = transceiverSendFrame(txMessage, txMessageSize, &txSlotsRemaining);

```

Comment:
I'm confused about how the txMessage receives information. It's defined as a local buffer in this function but would it not always just have nothing in it?

Reply:
This task will make a call to the Downlink Manager, passing along `txBuffer`. The downlink manager will then fill that buffer with a "ready to go" frame (serialized, etc.). The downlink manager isn't done yet, so for now yeah it will stay empty. The downlink manager is next up on Matthew's to-do list